src/wp-includes/class-wp-block-scanner.php

```diff
@@ -0,0 +1,1244 @@
+			 * allows for matching invalid JSON content.
+			 */
+			$comment_closing_at = strpos( $text, '-->', $json_at );
+			if ( false === $comment_closing_at ) {
+				$this->state      = static::COMPLETE;
+				$this->last_error = static::INCOMPLETE_INPUT;
+				return false;
+			}
+
+			/*
+			 * It looks like this logic leaves an error in here, when the position
+			 * overlaps the JSON or block name. However, for neither of those is it
+			 * possible to parse a valid block if that last overlapping character
+			 * is the void flag. This, therefore, will be valid regardless of how
+			 * the rest of the comment delimiter is written.
+			 */
+			if ( '/' === $text[ $comment_closing_at - 1 ] ) {
+				$has_void_flag    = true;
+				$void_flag_length = 1;
+			} else {
+				$has_void_flag    = false;
+				$void_flag_length = 0;
+			}
+
+			/*
+			 * If there's no JSON, then the span of text after the name
+			 * until the comment closing must be completely whitespace.
+			 */
+			if ( ! $has_json ) {
+				// This must be a block delimiter!
+				$this->state = static::MATCHED;
+				break;
+			}
+
+			// There's no JSON, so attempt to find its boundary.
+			$after_json_whitespace_length = 0;
+			for ( $char_at = $comment_closing_at - $void_flag_length - 1; $char_at > $json_at; $char_at-- ) {
+				$char = $text[ $char_at ];
+
+				switch ( $char ) {
+					case ' ':
+					case "\t":
+					case "\f":
+					case "\r":
+					case "\n":
+						++$after_json_whitespace_length;
+						continue 2;
+
+					case '}':
+						$json_length = $char_at - $json_at + 1;
+						break 2;
+
+					default:
+						++$at;
+						continue 3;
+				}
+			}
+
+			if ( 0 === $json_length || 0 === $after_json_whitespace_length ) {
+				$at = $this->find_html_comment_end( $comment_opening_at, $end );
+				continue;
+			}
+
+			// This must be a block delimiter!
+			$this->state = static::MATCHED;
+			break;
+		}
+
+		if ( static::MATCHED !== $this->state ) {
+			return false;
+		}
+
+		$this->next_token_at = $after_prev_delimiter;
+
+		$this->delimiter_at     = $comment_opening_at;
+		$this->delimiter_length = $comment_closing_at + 3 - $comment_opening_at;
+
+		$this->namespace_at     = $namespace_at;
+		$this->namespace_length = $namespace_length;
+
+		$this->name_at     = $name_at;
+		$this->name_length = $name_length;
+
+		$this->json_at     = $json_at;
+		$this->json_length = $json_length;
+
+		/*
+		 * When delimiters contain both the void flag and the closing flag
+		 * they shall be interpreted as void blocks, per the spec parser.
+		 */
+		$this->type = $has_void_flag
+			? static::VOID
+			: ( $has_closer ? static::CLOSER : static::OPENER );
+
+		$this->has_closing_flag = $has_closer;
+
+		if ( 'visit-html' === $html_spans && $comment_opening_at > $after_prev_delimiter ) {
+			$this->state = static::HTML_OPEN;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Constructor function.
+	 *
+	 * @since 6.9.0
+	 */
+	protected function __construct( string $source_text ) {
```

Comment:
Is there any reason this is protected? I see there is a create static function but the create static function appears to be doing the job this constructor is meant to do, albeit without the type safety and protections a constructor brings.

Could the doc block explain this or reference the correct way to construct a new scanner object? Most objects in WordPress use the `new` operator for object creation so this is a bit odd